app/workers/messaging.php

```diff
@@ -20,59 +30,180 @@
+    }
+
+    private function processMessage(Document $messageRecord, Document $providerRecord): void
+    {
+        $provider = match ($providerRecord->getAttribute('type')) {
+            'sms' => $this->sms($providerRecord),
+            'push' => $this->push($providerRecord),
+            'email' => $this->email($providerRecord),
+            default => null
+        };
+
+        $recipientsId = $messageRecord->getAttribute('to');
+
+        /**
+        * @var Document[] $recipients
+        */
+        $recipients = [];
+
+
+        $topics = $this->dbForProject->find('topics', [Query::equal('$id', $recipientsId)]);
+        foreach ($topics as $topic) {
+            $recipients = \array_merge($recipients, $topic->getAttribute('targets'));
         }
 
-        if (empty($this->from)) {
-            Console::info('Skipped sms processing. No phone number has been set.');
-            return;
+        $users = $this->dbForProject->find('users', [Query::equal('$id', $recipientsId)]);
+        foreach ($users as $user) {
+            $recipients = \array_merge($recipients, $user->getAttribute('targets'));
         }
 
-        $message = new SMS(
-            to: [$this->args['recipient']],
-            content: $this->args['message'],
-            from: $this->from,
-        );
+        $targets = $this->dbForProject->find('targets', [Query::equal('$id', $recipientsId)]);
+        $recipients = \array_merge($recipients, $targets);
 
-        try {
-            $this->sms->send($message);
-        } catch (\Exception $error) {
-            throw new Exception('Error sending message: ' . $error->getMessage(), 500);
+        $identifiers = \array_map(function (Document $recipient) {
+            return $recipient->getAttribute('identifier');
+        }, $recipients);
+
+        $maxBatchSize = $provider->getMaxMessagesPerRequest();
+        $batches = \array_chunk($identifiers, $maxBatchSize);
+        $deliveredTo = 0;
+
+        foreach ($batches as $batch) {
+            $messageRecord->setAttribute('to', $batch);
+            $message = match ($providerRecord->getAttribute('type')) {
+                'sms' => $this->buildSMSMessage($messageRecord, $providerRecord),
+                'push' => $this->buildPushMessage($messageRecord),
+                'email' => $this->buildEmailMessage($messageRecord, $providerRecord),
+                default => null
+            };
+            try {
+                $provider->send($message);
+                $deliveredTo += \count($batch);
+            } catch (Exception $e) {
+                $deliveryErrors = $messageRecord->getAttribute('deliveryErrors');
+                foreach ($batch as $identifier) {
+                    $deliveryErrors[] = 'Failed to send message to target' . $identifier . ': ' . $e->getMessage();
+                }
+                $messageRecord->setAttribute('deliveryErrors', $deliveryErrors);
+            }
         }
```

Comment:
Are we able to leverage coroutines here?

Reply:
using batch now